### PR TITLE
Code Review: Done

### PR DESCRIPTION
**Pull request by moonglum**
Hey Markus,

I just finished the code review of this gem. Everything's nice and dandy :+1:
However: I tweaked the README a little and adjusted some comments/documentation. So here is a little pull request. No code was changed.

Two questions:
## Nested class

In the Operation N-ary class you have a nested class. See [here](https://github.com/mbj/aql/blob/master/lib/aql/node/operation/nary.rb#L33-51). I've never seen that before. Why did you decide to go that route?
## Classes inheriting self

Why do some classes inherit `self`? I've never seen this before. What is the reasoning behind it?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,14 +17,51 @@ Installation
...
+```ruby
+gem 'aql', :git => 'https://github.com/mbj/aql'
+```
 
 Or use ```:github => 'mbj/aql'``` if you prefer unencrypted protocols.
 
 Examples
 --------
 
-No public api => No examples. (Subjected to change).
```
Comment (mbj) [inline README.md:27]: Yeah we need a DSL ;). Thx for doing my job here.

**Comment by snusnu** [comment]
@moonglum inheriting a class from `self` is mostly a question of style. Since `self` points to the enclosing class, it's exactly the same as naming that constant explicitly. The only difference is that it's shorter, basically acts like accessing a variable instead of "the real thing", and thus is more robust wrt class rename refactorings. Note that the last argument only holds as long as the class doesn't get moved out of its namespace.

Re private classes, there's no such thing in ruby. It doesn't matter if the class definition "is in a private section". It's just another regular (public) class.

**Comment by mbj** [comment]
@moonglum Thx for these doc fixes. You get blind over the time when heavily digging code ;)

/re nested  class: It directly reflects the fact AQL has a "COLLECT" and a "COLLECT INTO statement". 

@snusnu, thx for explanation, you just saved me some typing.

**Comment by moonglum** [comment]
@mbj No problem ;) Yep, I know that problem ;)

Nested Classes:
Ok, I understand. It's quite unusual for me, but I understand the reasoning :)

Inheriting self:
@snusnu Ok! That makes sense :) Thanks for the explanation!

"Private classes":
Yep, I know that ;) I just referred to an occurrence in the code, where you had the private keyword on one line and the class directly beneath. That looked kind of odd to me ;)

**Comment by mbj** [comment]
@moonglum The whole class structure is optimized for the minimum amount of mutations to kill. Which is IMHO the best metric you can optimize. For this reason the COLLECT INTO statement, is just a special case of the COLLECT statement. //cc @dkubb

The class structure is deep, but it reflects the code inheritance graph. I generally only inherit for code reuse never because there is such a "kind of" stuff you learn at university. If there is a code inheritance, there is also enough "kind of behavior".

**Comment by dkubb** [comment]
> The class structure is deep, but it reflects the code inheritance graph. I generally only inherit for code reuse never because there is such a "kind of" stuff you learn at university. If there is a code inheritance, there is also enough "kind of behavior".

That's interesting because I inherit when my intention that an object be used as a "stand in" for instances of an ancestor. For code sharing between unrelated objects I would tend to use composition or modules.

**Comment by mbj** [comment]
@dkubb Hehe, I thought I'd need your voice to clarify my thoughts here. Hence the cc.

I also favor composition. I think we are in sync here. So I have to find a better description. I think there is a difference in behavior and state composition. If you compose behavior your subclass, of you compose state you use composition. This should fit more.

As all nodes share behavior so the subclass is the tool of choice. The tool with minimal power to do the job.

I usually NOT subclass to add state. The base class Node does not have any state. All direct non abstract subclasses add one field, and there are only rare edge cases where a 3rd level subclass adds a field.

Hopefully this describes my reasoning better. Summary I still think the design is good ;)

**Comment by mbj** [comment]
I just notice this one is enough:

 "I use inheritance when the public interface does not change (Liskov Substitution Principle), and it avoids possible code duplication".

I have no problem with the fact the AST here is a deeply nested structure, the public interface is `#visit` and `#to_aql`. 

Dunno why I created such a lengthy discussion ;)

**Pull request merged**